File: scripts/gravar.js
#!/usr/bin/env gjs

const GLib = imports.gi.GLib;
const Gio = imports.gi.Gio;
const System = imports.system;

// Tenta importar a biblioteca moderna
let GLibUnix = null;
try {
    GLibUnix = imports.gi.GLibUnix;
} catch (e) {}

// Conectar ao DBus
const bus = Gio.bus_get_sync(Gio.BusType.SESSION, null);

const proxy = Gio.DBusProxy.new_sync(
    bus,
    Gio.DBusProxyFlags.NONE,
    null,
    "org.gnome.Shell.Screencast",
    "/org/gnome/Shell/Screencast",
    "org.gnome.Shell.Screencast",
    null
);

print("A iniciar gravação...");

try {
    proxy.call_sync(
        "Screencast",
        new GLib.Variant("(sa{sv})", ["", {}]), 
        Gio.DBusCallFlags.NONE,
        -1,
        null
    );
    print("Gravação a correr (PID: " + System.programInvocationName + ")...");
} catch (e) {
    print("Erro ao iniciar: " + e.message);
    System.exit(1);
}

const loop = new GLib.MainLoop(null, false);

function stop() {
    try {
        print("A parar...");
        proxy.call_sync("StopScreencast", null, Gio.DBusCallFlags.NONE, -1, null);
    } catch (e) {}
    loop.quit();
    return GLib.SOURCE_REMOVE;
}

// === LÓGICA DE SINAIS CORRIGIDA ===
function adicionar_sinal(sinal, callback) {
    // Tenta o método novo (SEM PRIORIDADE) - Ubuntu 22.04/24.04
    if (GLibUnix && typeof GLibUnix.add === 'function') {
        try {
            return GLibUnix.add(sinal, callback);
        } catch (e) {}
    }
    
    // Tenta o método antigo (COM PRIORIDADE) - Fallback
    try {
        return GLib.unix_signal_add(GLib.PRIORITY_DEFAULT, sinal, callback);
    } catch (e) {
        // Se ambos falharem, não faz nada (o script corre mas só fecha com kill -9)
    }
}

// 15 = SIGTERM, 2 = SIGINT
adicionar_sinal(15, stop);
adicionar_sinal(2, stop);

loop.run();

File: scripts/gravar.py
#!/usr/bin/env python3
import sys
import signal
import gi
import os

# Importar bibliotecas do GNOME
gi.require_version('Gio', '2.0')
gi.require_version('GLib', '2.0')
from gi.repository import Gio, GLib

# Configurações do DBus
BUS_NAME = "org.gnome.Shell.Screencast"
OBJECT_PATH = "/org/gnome/Shell/Screencast"
INTERFACE_NAME = "org.gnome.Shell.Screencast"

def main():
    # 1. Conectar ao DBus da Sessão
    bus = Gio.bus_get_sync(Gio.BusType.SESSION, None)
    
    proxy = Gio.DBusProxy.new_sync(
        bus,
        Gio.DBusProxyFlags.NONE,
        None,
        BUS_NAME,
        OBJECT_PATH,
        INTERFACE_NAME,
        None
    )

    # 2. Definir função de paragem
    loop = GLib.MainLoop()

    def stop_recording(signum, frame):
        try:
            print("A parar gravação...", flush=True)
            proxy.call_sync(
                "StopScreencast",
                None,
                Gio.DBusCallFlags.NONE,
                -1,
                None
            )
            print("Guardado em ~/Vídeos.", flush=True)
        except Exception as e:
            print(f"Nota ao parar: {e}", flush=True)
        
        loop.quit()

    # 3. Capturar sinais (Ctrl+C ou kill)
    signal.signal(signal.SIGTERM, stop_recording)
    signal.signal(signal.SIGINT, stop_recording)

    # 4. Iniciar Gravação
    print("A iniciar...", flush=True)
    try:
        # O segredo: Passar GLib.Variant explícito para não haver erro de tipos
        # (sa{sv}) -> String (template), Array de Dicionário (opções)
        params = GLib.Variant('(sa{sv})', ("", {}))
        
        proxy.call_sync(
            "Screencast",
            params,
            Gio.DBusCallFlags.NONE,
            -1,
            None
        )
        print("Gravação ativa.", flush=True)
    except Exception as e:
        print(f"ERRO FATAL: {e}", flush=True)
        sys.exit(1)

    # 5. Manter script vivo
    loop.run()

if __name__ == "__main__":
    main()
